### PR TITLE
[Cling] ROOT-6967 Add direct way too lookup tags and put it to a good use in TCling

### PR DESCRIPTION
**Pull request by dpiparo**
and put it to a good use in Cling

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build

File: interpreter/cling/lib/Utils/AST.cpp [hunk 1]
```diff
@@ -49,6 +49,9 @@ namespace utils {
   NestedNameSpecifier* GetFullyQualifiedNameSpecifier(const ASTContext& Ctx,
                                                       NestedNameSpecifier* scope);
 
+  template<typename D>
+  static D* LookupResult2Decl(LookupResult& R);
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Utils/AST.cpp]: Can you remove this and move the definition into an unnamed namespace?

File: interpreter/cling/lib/Utils/AST.cpp [hunk 2]
```diff
@@ -1441,14 +1457,26 @@ namespace utils {
...
                    Sema::ForRedeclaration);
     Lookup::Named(S, R, Within);
-    if (R.empty())
-      return 0;
+    return LookupResult2Decl<clang::NamedDecl>(R);
+  }
 
-    R.resolveKind();
+  TagDecl* Lookup::Tag(Sema* S, llvm::StringRef Name,
+                           const DeclContext* Within) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Utils/AST.cpp]: Indentation

File: interpreter/cling/lib/Utils/AST.cpp [hunk 3]
```diff
@@ -1441,14 +1457,26 @@ namespace utils {
...
+    DeclarationName DName = &S->Context.Idents.get(Name);
+    return Lookup::Tag(S, DName, Within);
+  }
 
-    if (R.isSingleResult())
-      return R.getFoundDecl();
-    return (clang::NamedDecl*)-1;
+  TagDecl* Lookup::Tag(Sema* S, const char* Name,
+                           const DeclContext* Within) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Utils/AST.cpp]: Indentation

File: interpreter/cling/lib/Utils/AST.cpp [hunk 4]
```diff
@@ -1441,14 +1457,26 @@ namespace utils {
...
+    return Lookup::Tag(S, llvm::StringRef(Name), Within);
+  }
+
+  TagDecl* Lookup::Tag(Sema* S, const clang::DeclarationName& Name,
+               const DeclContext* Within) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Utils/AST.cpp]: Indentation

File: interpreter/cling/lib/Utils/AST.cpp [hunk 5]
```diff
@@ -1441,14 +1457,26 @@ namespace utils {
...
+    LookupResult R(*S, Name, SourceLocation(), Sema::LookupTagName,
+                   Sema::ForRedeclaration);
+    Lookup::Named(S, R, Within);
+    return LookupResult2Decl<clang::TagDecl>(R);
```
Comment (vgvassilev) [inline interpreter/cling/lib/Utils/AST.cpp]: The clang::LookupResult supports filters. They can likely help and you could remove or simplify that routine. We use filters somewhere in TCling, else just grep for LookupFilter.

Comment (Axel-Naumann) [reply interpreter/cling/lib/Utils/AST.cpp]: `git grep LookupFilter -- core/metacling/src/` is empty. Also I'm not sure what to filter on here? You mean to only find `TagDecl`s?

File: interpreter/cling/include/cling/Utils/AST.h [hunk 6]
```diff
@@ -238,6 +238,45 @@ namespace utils {
...
+    ///\brief Quick lookup for a single named tag declaration ( enums, classes,
+    /// structs, and unions) in a given declaration context.
+    ///
+    ///\param[in] S - Semantic Analysis object doing the lookup.
+    ///\param[in] Name - The name we are looking up.
+    ///\param[in] Within - The context within the lookup is done. If 0 the
+    ///                    TranslationUnitDecl is used.
+    ///\returns the found result if single, -1 if multiple or 0 if not found.
+    ///
+    clang::TagDecl* Tag(clang::Sema* S,
```
[the remark this answers is not in the document]
Comment (dpiparo) [reply interpreter/cling/include/cling/Utils/AST.h]: Thanks for the suggestion: absolutely.

File: interpreter/cling/lib/Utils/AST.cpp [hunk 7]
```diff
@@ -1425,6 +1428,19 @@ namespace utils {
     return dyn_cast<NamespaceDecl>(R.getFoundDecl());
   }
 
+  template<typename D>
+  static D* LookupResult2Decl(LookupResult& R)
```
[the remark this answers is not in the document]
Comment (dpiparo) [reply interpreter/cling/lib/Utils/AST.cpp]: Yes of course.

File: interpreter/cling/lib/Utils/AST.cpp (hunk 1 above)
Comment (dpiparo) [reply interpreter/cling/lib/Utils/AST.cpp]: Yes of course.

File: interpreter/cling/lib/Utils/AST.cpp (hunk 2 above)
Comment (dpiparo) [reply interpreter/cling/lib/Utils/AST.cpp]: thanks. Can I use clang-format to check it thoroughly? I _think_ it's different from ROOT's one.

File: interpreter/cling/lib/Utils/AST.cpp (hunk 3 above)
Comment (dpiparo) [reply interpreter/cling/lib/Utils/AST.cpp]: apologies. See above.

File: interpreter/cling/lib/Utils/AST.cpp (hunk 4 above)
Comment (dpiparo) [reply interpreter/cling/lib/Utils/AST.cpp]: ok.

**Comment by dpiparo** [comment]
all comments integrated.

File: interpreter/cling/test/Lookup/tag.C [hunk 8]
```diff
@@ -0,0 +1,77 @@
+//------------------------------------------------------------------------------
```
Comment (Axel-Naumann) [inline interpreter/cling/test/Lookup/tag.C]: Wow, thanks!

Comment (dpiparo) [reply interpreter/cling/test/Lookup/tag.C]: thanks for the feedback.

**Pull request merged**